pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1156 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ipam
+
+import (
+	"container/heap"
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"net"
+	"sync"
+	"time"
+
+	"k8s.io/api/core/v1"
+	networkingv1alpha1 "k8s.io/api/networking/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	informers "k8s.io/client-go/informers/core/v1"
+	networkinginformers "k8s.io/client-go/informers/networking/v1alpha1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	networkinglisters "k8s.io/client-go/listers/networking/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/component-base/metrics/prometheus/ratelimiter"
+	nodeutil "k8s.io/component-helpers/node/util"
+	"k8s.io/klog/v2"
+	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	cidrset "k8s.io/kubernetes/pkg/controller/nodeipam/ipam/multicidrset"
+	controllerutil "k8s.io/kubernetes/pkg/controller/util/node"
+	"k8s.io/kubernetes/pkg/util/slice"
+	netutil "k8s.io/utils/net"
+)
+
+const (
+	defaultClusterCIDRKey        = "kubernetes.io/clusterCIDR"
+	defaultClusterCIDRValue      = "default"
+	defaultClusterCIDRName       = "default-cluster-cidr"
+	defaultClusterCIDRAPIVersion = "networking.k8s.io/v1alpha1"
+	clusterCIDRFinalizer         = "networking.k8s.io/cluster-cidr-finalizer"
+	ipv4MaxCIDRMask              = 32
+	ipv6MaxCIDRMask              = 128
+	minPerNodeHostBits           = 4
+	maxRetries                   = 15
+)
+
+// raWaitGroup is used to wait for the RangeAllocator to finish the goroutines.
+var raWaitGroup sync.WaitGroup
```

Comment:
We shouldn't use global variables for this.

Reply:
Made it local